tket/src/Circuit/Multiplexor.cpp

```diff
@@ -630,29 +643,222 @@ void MultiplexedU2Box::generate_circuit() const {
+  ctrl_tensored_op_map_t new_op_map;
+  for (auto it = op_map_.begin(); it != op_map_.end(); it++) {
+    std::vector<Op_ptr> ops;
+    for (auto op : it->second) {
+      ops.push_back(op->transpose());
+    }
+    new_op_map.insert({it->first, ops});
+  }
+  return std::make_shared<MultiplexedTensoredU2Box>(new_op_map, impl_diag_);
+}
+
+op_signature_t MultiplexedTensoredU2Box::get_signature() const {
+  op_signature_t qubits(n_controls_ + n_targets_, EdgeType::Quantum);
+  return qubits;
+}
+
+nlohmann::json MultiplexedTensoredU2Box::to_json(const Op_ptr &op) {
+  const auto &box = static_cast<const MultiplexedTensoredU2Box &>(*op);
+  nlohmann::json j = core_box_json(box);
+  j["op_map"] = box.get_op_map();
+  j["impl_diag"] = box.get_impl_diag();
+  return j;
+}
+
+Op_ptr MultiplexedTensoredU2Box::from_json(const nlohmann::json &j) {
+  MultiplexedTensoredU2Box box = MultiplexedTensoredU2Box(
+      j.at("op_map").get<ctrl_tensored_op_map_t>(),
+      j.at("impl_diag").get<bool>());
+  return set_box_id(
+      box,
+      boost::lexical_cast<boost::uuids::uuid>(j.at("id").get<std::string>()));
+}
+
+void MultiplexedTensoredU2Box::generate_circuit() const {
+  Circuit circ(n_controls_ + n_targets_);
+  // contains the multiplexed-Rz gates
+  Circuit diag_circ(n_controls_ + n_targets_);
+  // the final diagonal on the control qubits
+  Eigen::VectorXcd diag_vec =
+      Eigen::VectorXcd::Constant(1ULL << n_controls_, 1);
+  for (unsigned i = 0; i < n_targets_; i++) {
+    ctrl_op_map_t u2_op_map;
+    for (auto it = op_map_.begin(); it != op_map_.end(); it++) {
+      u2_op_map.insert({it->first, it->second[i]});
+    }
+    MultiplexedU2Box mbox(u2_op_map);
+    Circuit inner_circ;
+    Eigen::VectorXcd inner_diag_vec;
+    std::tie(inner_circ, inner_diag_vec) = mbox.decompose();
+    std::vector<unsigned> args(n_controls_);
+    std::iota(std::begin(args), std::end(args), 0);
+    args.push_back(i + n_controls_);
+    // append the first part of the decomposition
+    circ.append_qubits(inner_circ, args);
+    // disentangle one qubit from the diagonal
+    // results in a multiplexed-Rz targeting the target j
+    ctrl_op_map_t multip_rz;
+    for (unsigned long long j = 0; j < (1ULL << n_controls_); j++) {
+      Complex a = inner_diag_vec[2 * j];
+      Complex b = inner_diag_vec[2 * j + 1];
+      // convert diag[a,b] into a p*Rz(alpha)
+      double a_phase = std::arg(a);
+      double b_phase = std::arg(b);
+      double alpha = (b_phase - a_phase) / PI;
+      Complex p = std::exp((b_phase + a_phase) * 0.5 * i_);
+      std::vector<bool> bitstr = dec_to_bin(j, n_controls_);
+      if (std::abs(alpha) > EPS) {
+        multip_rz.insert({bitstr, get_op_ptr(OpType::Rz, alpha)});
+      }
+      // update the diagonal on the control qubits
+      diag_vec[j] *= p;
+    }
+    if (!multip_rz.empty()) {
+      diag_circ.add_box(MultiplexedRotationBox(multip_rz), args);
+    }
+  }
+  if (impl_diag_) {
```

Comment:
So if `impl_diag_=False` then the Diagonal operator isn't added to the Circuit at all?

I'd maybe suggest that the functionality of `impl_diag_` is as to whether the diagonal operator is decomposed in this `generate_circuit` or left as a `DiagonalBox`. Ideally if a user chooses to leave it as a `DiagonalBox` then there are later compilation passes they can apply (that we would need to write!) that would suitably commute these through end of circuit measures, or other sets of controls etc to lead to gate cancellations.

Reply:
Makes sense. I would suggest that the behaviour of `impl_diag_` in `MultiplexedU2Box` should also be changed for the sake of consistency. It sounds like the commute DiagonalBox pass you suggested must be prioritised because removing them is not straightforward with the python API.

Reply:
Agreed on consistency suggestion.

Also agreed on prioritising such a commuting `DiagonalBox` pass. We could maybe do something sneaky by

Reply:
On second thought, since the decomposition already produces box type gates, I don't see any reason to decompose the `DiagonalBox` in `generate_circuit`. How about I simply remove the `impl_diag_` from `MultiplexedTensorU2Box`? We can keep `MultiplexedU2Box` as it is for now.

Reply:
Works for me!